pkg/shed/index.go

```diff
@@ -303,21 +310,68 @@ func (f Index) Iterate(fn IndexIterFunc, options *IterateOptions) (err error) {
 			return fmt.Errorf("encode key: %w", err)
 		}
 	}
+
 	it := f.db.NewIterator()
 	defer it.Release()
 
+	var ok bool
+
 	// move the cursor to the start key
-	ok := it.Seek(startKey)
-	if !ok {
-		// stop iterator if seek has failed
-		return it.Error()
+	ok = it.Seek(startKey)
+
+	if !options.Reverse {
+		if !ok {
+			// stop iterator if seek has failed
+			return it.Error()
+		}
+	} else {
+		// reverse seeker
+		if options.StartFrom != nil {
```

Comment:
lots of nesting here. can't you just check that `!ok` just after the seek? if the original seek failed the whole function should return anyway no?

Reply:
Excuse this commit, I figured that I have left enough comments for any reader to follow logic.

This change revolves around adding support for reverse iteration over items in index, which are stored (in this case) in LevelDB (but changes should be same for many other similar style KV databases). Change being made is not to just support reverse iteration, but to add support for it, while also supporting "standard" iteration (from first key in index onward). Before first seek no check is performed to know exactly what is contained in the `startKey`.

Because there are multiple possible conditions, case in which error should checked must exists on multiple places. It may be possible to have more complex single condition first, however I think due to the nature of conditions second must be present also. I might have not considered some condition, but I see no way to make it while also leaving code in "readable" condition. I am open for suggestions.

First seek to `startKey` is left from old code, and it is fine to exists (even though currently not used for new 'reverse' use-case, it might be used on some other occasion if the `options.StartFrom` is configured with `options.Reverse`). That `startKey` could either be just index prefix, index prefix with some prefix, or some of those prefix combinations with exact key that should be found.

First `if` condition is to to check if we are iterating from topmost item from LevelDB (at least with prefix), and failing if there was error in that case. Conditions `!options.Reverse` and `!ok` here cannot be joined, as we need `else` condition for the case where 'reverse' iteration is performed. Of course we could have separate condition with both `if !options.Reverse && !ok {...`, and then after that another `if options.Reverse {...`, but those are functionally same (or have some couple of other complex conditions).

At this point we are in the case for reverse order. Separate check if `options.StartFrom` is configured is needed in this case because exact key should be found in any case.

In the case where we only have prefix, and are doing reverse iteration, a bit more complex operation for finding end is needed. This is especially true because this index structure can contain multiple indexes in the same LevelDB database.

Initial step is to just try to find last item in the database, using `it.Last()`. In the case where there is only single index, we have last item, and are done there. However, when there are multiple indexes we have additional work to find last item for needed index (and prefix).

This is done in the `if lastKeyHasPrefix := bytes.HasPrefix(it.Key(), prefix); !lastKeyHasPrefix {...` condition. This is only performed in the case when `options.StartFrom` is not configured.
Way to find last item in index (for prefix) is to find first next item, and then go to previous item, at which point cursor should be on valid item. Note here that I also partly rely on later logic which checks if item has proper prefix before returning item.

In the case where multiple indexes exists, or we also have some prefix, finding key after some other is (sort-of) known. Since keys are basically sorted byte arrays, incrementing last byte in array for key and seeking to it should move to subsequent key, even if exact key that was put in the `Seek` function does not exists.

For this example (for index `\x02` - first index ID, and prefix `want`):

```txt
"\x00" "{\"fields\":{},\"indexes\":{\"2\":{\"name\":\"retrieval-index\"},\"3\":{\"name\":\"second-index\"}}}
"\x02skip-hash-01" "value"
"\x02skip-hash-02" "value"
"\x02want-hash-00" "value"
"\x02want-hash-04" "value"
"\x02want-hash-05" "value"
"\x03iterate-hash-10" "value"
```

we want to iterate over items with prefix `\x02want`, incremented prefix will be `\x02wanu`, which will on `Seek` operation move cursor to `\x03iterate-hash-10`. Moving to previous database item should move cursor to correct index.
For this case also we could use new prefix of `\x02wao` (if the actual prefix was `\x02wan\xff`), and which would result in the same end result. In the case where only index prefix is used, i.e. prefix is `\x02`, the new prefix would be `\x03`, and same end goal would be reached.

In the case where last byte of the key is `0xFF` (which is, of course, highly unlikely or impossible in our case) we remove that last byte, and increment previous one.

I do not consider this in any way brute-force way, there are only 2 operations here: one to find next item, and one to go backward.
This logic cannot be compared to something like in PR #821, where there actual numerical offset is used to retrieve page. Even though actual logic is correct, which is why I approved it, it is a bit too much. Not sure why actual pages are used, and why have we not returned some cursor, which could be used for further iterations. Or even last pin could be used to start from that one.

The tests I added (copied and changed) should cover most conditions pretty well.
I am open to precise suggestion on how current change can be improved.

Reply:
Cool. Thanks for the explanation. If you have suggestions to improve other pull request you should make them :+1: